### PR TITLE
[Form] DateIntervalType: Allow to configure labels & enhance form theme

### PR DESCRIPTION
**Pull request by ogizanagi**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no (unless someone relies on this non themed type)
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | Should document the new `labels` option

I just realized by using it for last fixes in #20886 and #20877 that this type was not really themed:

### before

screenshot 2016-12-13 a 00 54 35

At least labels should appear, but this also means being able to change them (thus the new `labels` option).

I think the form themes should provide a functional & minimalistic integration like this:

### after

screenshot 2016-12-13 a 00 54 17

---
(On screenshots above, I've only added a css rule to remove the 100% width of the `.table` class. See https://github.com/symfony/symfony/pull/20887#discussion_r92069557)

### CODE REVIEW

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_layout.html.twig [hunk 1]
```diff
@@ -94,13 +94,32 @@
...
             {{- form_errors(form) -}}
-            {%- if with_years %}{{ form_widget(form.years) }}{% endif -%}
-            {%- if with_months %}{{ form_widget(form.months) }}{% endif -%}
-            {%- if with_weeks %}{{ form_widget(form.weeks) }}{% endif -%}
-            {%- if with_days %}{{ form_widget(form.days) }}{% endif -%}
-            {%- if with_hours %}{{ form_widget(form.hours) }}{% endif -%}
-            {%- if with_minutes %}{{ form_widget(form.minutes) }}{% endif -%}
-            {%- if with_seconds %}{{ form_widget(form.seconds) }}{% endif -%}
+            <div class="table-responsive">
+                <table class="table table-bordered table-condensed table-striped" style="width: auto;">
```
Comment (ogizanagi) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_layout.html.twig]: `style="width: auto;` should be removed I guess, however the default bootstrap `.table` class adds a `width: 100%` rule. But we should probably don't care and let the user apply some additional style?

**Comment by fabpot** [comment]
I think nobody can understand the `invert` part, should we remove it?

**Comment by ogizanagi** [comment]
Good question. I don't know. It can be explained better in a real application by adding a help block describing its purpose.
However I do think the feature behind makes sense.

**Comment by fabpot** [comment]
At least, we should make the text understandable by default (`invert` does not say anything about what it does). `Negative interval` does not sounds good. It's more about an `Interval in the past`? Even the main `Interval` label is not good enough if you ask me. Let's try to find good names here.

**Comment by ogizanagi** [comment]
I agree. `Interval is in the past` is better.
About the main `Interval` label, it's only the name of the field in my form.

I know @javiereguiluz is really gifted at suggesting good names for such things and express good opinions. Maybe he can help 😃

**Comment by javiereguiluz** [comment]
It's tricky to find a good name for this option. In [the Wikipedia page](https://en.wikipedia.org/wiki/ISO_8601) they actually use "negative" (mostly as "negative offset").

Other words that I looked up were: "reverse interval" and "backwards interval". The first one seems to be more popular.

---

I don't like *"Interval is in the past"* because I understand it as follows:

* A "negative 3 days interval" means "go back in time 3 days from now".
* A "3 days interval in the past" means "go back 3 days ... and do that some undefined time in the past".

---

So let's ping some English native contributors to help us: @weaverryan @michaelcullum  Thanks!

**Comment by michaelcullum** [comment]
I'd be in agreement with Javier about the interpretation of `3-day negative interval` versus `3-day interval in the past` (the latter meaning that defined interval at some point in the past as opposed to going back by that defined interval).

The main problem here is the word interval as an interval has no direction/is a scalar; therefore an interval is always positive (e.g. in music intervals it's always called an octave, whether it's up an octave or down an octave). It's not the responsibility of the interval to store if it should be negative (inverted) or not, but of when you're putting it relative to something to work something out (e.g. working out a start from an end). Therefore the interval is not negated/reversed, but the interval is being subtracted from a point; it's tricky to find a name because we're basically naming the wrong thing.

That said, backwards interval, negative interval and reverse interval all make sense to me although in my opinion, a negative interval would make the most amount of sense.

**Comment by ogizanagi** [comment]
Many thanks to both of you @javiereguiluz and @michaelcullum. 🙏 
As it seems the "Negative interval" does not look that bad to you, I've changed the default value for this for now.

**Comment by ro0NL** [comment]
What about "Time interval" and "Subtract interval?"

?

File: src/Symfony/Component/Form/Extension/Core/Type/DateIntervalType.php [hunk 2]
```diff
@@ -192,6 +195,21 @@ public function configureOptions(OptionsResolver $resolver)
...
+            return array_replace(array(
+                'years' => null,
+                'months' => null,
+                'days' => null,
+                'weeks' => null,
+                'hours' => null,
+                'minutes' => null,
+                'seconds' => null,
+                'invert' => 'Negative interval',
+            ), array_filter($labels, function ($label) {
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/DateIntervalType.php]: This can be simplified by `array_filter($labels)`.

Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Core/Type/DateIntervalType.php]: No, because it'll filter any empty value.
It case someone simply want to disable the label, he'll set it to `false`. But `false` would be removed and thus impossible to disable the label rendering from options.

Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Core/Type/DateIntervalType.php]: @HeahDude : Do you think gathering the labels under a single option is a good idea or should I opt back for `label_x` options?

#21002 is suggesting to translate `DateTimeType` subtypes labels and uses dedicated options. I guess it's coherent regarding other options actually, even if it'll looks ugly in the case of the `DateIntervalType`. But I guess we should be consistent. WDYT?

Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/DateIntervalType.php]: I prefer your approach :). Problem is that it mostly depends on the widget we consider.

Having the possibility to define labels is great, mainly to translate them. It means we should be able to define all of them (and that would be consistent).

Also regarding performances, your approach should be better, using one normalizer for one array (even calling an `array_replace` and an `array_filter`) instead of defining many options. Resolving only one option involves cloning the resolver instance and many calls, so that must cost much more.

In conclusion, I'd say maybe #20002 should follow what's implemented here, using `date_label` and `time_label` as string or array, so maybe `DateType` and `TimeType` should have a `labels` option too.

What do you and others think?

Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Core/Type/DateIntervalType.php]: I opted for this approach as it was my own preference, but expected negative feedbacks about it. Apparently it doesn't look that bad, so I'm satisfied 😅 .

Apart from perfs, IMHO segregating similar options is cluttering form types and make documentation more complicated and verbose no gain. Of course it depends of use cases and options, as it makes sense for most of them to be separated from each others in order to benefits of fine control over OptionResolver features like validation and normalization. But even for some of them, a feature like #18134 would be interesting (with an extra cost, though).

My only fear was about consistency with other form types which opted for segregation and autocompletions with the Symfony plugin for PHPStorm for instance.

**Comment by ogizanagi** [comment]
> What about "Time interval" and "Subtract interval?"

Thanks @ro0NL :) . 
As a non native english speaker, I still prefer "Negative interval" though, as it seems to be approved.

As there is no new wording propositions that would be appropriate, I think this is ready (Travis failure unrelated. Can be rebased though).

**Comment by fabpot** [comment]
Thank you @ogizanagi.
